src/ModelContextProtocol/Protocol/Messages/ProgressNotification.cs

```diff
@@ -0,0 +1,112 @@
+﻿using System.ComponentModel;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace ModelContextProtocol.Protocol.Messages;
+
+/// <summary>
+/// An out-of-band notification used to inform the receiver of a progress update for a long-running request.
+/// <see href="https://github.com/modelcontextprotocol/specification/blob/main/schema/">See the schema for details</see>
+/// </summary>
+[JsonConverter(typeof(Converter))]
+public class ProgressNotification
+{
+    /// <summary>
+    /// The progress token which was given in the initial request, used to associate this notification with the request that is proceeding.
+    /// </summary>
+    public required ProgressToken ProgressToken { get; init; }
+
+    /// <summary>
+    /// The progress thus far. This should increase every time progress is made, even if the total is unknown.
+    /// </summary>
+    public required ProgressNotificationValue Progress { get; init; }
+
+    /// <summary>Provides a <see cref="JsonConverter"/> for <see cref="ProgressNotification"/>.</summary>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public sealed class Converter : JsonConverter<ProgressNotification>
+    {
+        /// <inheritdoc />
+        public override ProgressNotification? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            ProgressToken? progressToken = null;
+            float? progress = null;
+            float? total = null;
+            string? message = null;
+            
+            while (reader.Read() && reader.TokenType != JsonTokenType.EndObject)
+            {
+                if (reader.TokenType == JsonTokenType.PropertyName)
+                {
+                    var propertyName = reader.GetString();
+                    reader.Read();
+                    switch (propertyName)
+                    {
+                        case "progressToken":
+                            if (JsonSerializer.Deserialize(ref reader, options.GetTypeInfo(typeof(ProgressToken))) is not ProgressToken token)
+                            {
+                                throw new JsonException("Invalid value for 'progressToken'.");
+                            }
```

Comment:
This path is never going to be hit, so we can simplify:
```suggestion
                            ProgressToken token = JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<ProgressToken>());
```

Reply:
Also, we should replace redundant `options.GetTypeInfo<T>()` calls with direct referencing of JTOs supplied by the source generated context. I'm already in the process of eliminating most of those in my local changes.

Reply:
> we should replace redundant options.GetTypeInfo<T>() calls with direct referencing of JTOs supplied by the source generated context

And ignore the passed in `options`? That's "ok"?